gibbon-compiler/tests/HoistBoundsCheck.hs

```diff
@@ -0,0 +1,307 @@
+                (EndOfReg (SingleR "r1") Output (SingleR "end_r1"))
+          )
+        $ LetE ("b", [], MkTy2 BoolTy, LitE 1)
+        $ IfE
+          (VarE "b")
+          (MkProdE [])
+          (MkProdE [])
+
+
+case_t2 :: Assertion
+case_t2 = assertBool "Output matches one of the expected variants" $ actual `elem` [expected1, expected2, expected3, expected4]
+  where
+    actual = fst $ defaultPackedRunPassM $ hoistBoundsCheck test2 S.empty
+
+    test2 :: L2.Exp2
+    test2 =
+      LetE ("b", [], MkTy2 BoolTy, LitE 1) $
+        IfE
+          (VarE "b")
+          ( Ext
+              $ LetRegionE (VarR "r1") Undefined Nothing
+              $ LetE
+                ( "_",
+                  [],
+                  MkTy2 IntTy,
+                  Ext $
+                    BoundsCheckVector
+                      [
+                        (
+                          10,
+                          (Loc (LREM (singleLocVar "l1") (SingleR "r1") (SingleR "end_r1") Output)),
+                          (EndOfReg (SingleR "r1") Output (SingleR "end_r1"))
+                        ) , 
+                        (
+                          10,
+                          (Loc (LREM (singleLocVar "l2") (SingleR "r2") (SingleR "end_r2") Output)),
+                          (EndOfReg (SingleR "r2") Output (SingleR "end_r2"))
+                        )
+                      ]
+                )
+              $ Ext
+              $ LetLocE (singleLocVar "l1") (StartOfRegionLE (VarR "r1"))
+              $ Ext 
+              $ LetRegionE (VarR "r2") Undefined Nothing
+              $ Ext
+              $ LetLocE (singleLocVar "l2") (StartOfRegionLE (VarR "r2"))
+              $ (MkProdE [])
+          )
+          (MkProdE [])
+
+    expected1 :: L2.Exp2
+    expected1 =
+      Ext
+        $ LetRegionE (VarR "r2") Undefined Nothing
+        $ Ext
+        $ LetRegionE (VarR "r1") Undefined Nothing
+        $ Ext
+        $ LetLocE (singleLocVar "l1") (StartOfRegionLE (VarR "r1"))
+        $ Ext 
+        $ LetLocE (singleLocVar "l2") (StartOfRegionLE (VarR "r2"))
+        $ LetE
+          ( "_",
+            [],
+            MkTy2 IntTy,
+            Ext $
+              BoundsCheckVector 
+              [
+                (
+                  10,
+                  (Loc (LREM (singleLocVar "l1") (SingleR "r1") (SingleR "end_r1") Output)),
+                  (EndOfReg (SingleR "r1") Output (SingleR "end_r1"))
+                )
+                ,
+                (
+                  10,
+                  (Loc (LREM (singleLocVar "l2") (SingleR "r2") (SingleR "end_r2") Output)),
+                  (EndOfReg (SingleR "r2") Output (SingleR "end_r2"))
+                )
+              ]
+          )
+        $ LetE ("b", [], MkTy2 BoolTy, LitE 1)
+        $ IfE
+          (VarE "b")
+          (MkProdE [])
+          (MkProdE [])
+
+    --multiple correct answers here 
```

Comment:
How come there are several correct answers? Is there anything nondeterministic about the algorithm? This needs an explanation.

Reply:
Haskell maps and sets are unordered data structures. When you do folds and maps over these structures the order in which you inserted elements is not guaranteed. The algorithm does not depend on the order, but when you do a fold, parallel dependencies may have different order on different instances/machines etc. This is fine. 
Hence the multiple correct answers.

Reply:
If you try to protect from any possible changes to out Ord instances in the future, you have to provide some argument why you think we'll be changing them. Also, defending against multitudes of possible futures is not something that any project I know of does.

Reply:
We can't protect ourselves from all posible futures. Sure, in the next version GHC can remove `Ord` altogether, how do we protect against that?..

Also, you seem to be misunderstanding how type classes work: GHC can only change the definition of the Ord class (very unlikely because it's a part of the Haskell standard that has not changed in decades), but not the actual comparison predicate, which is defined in our own code in the `Ord` **instance**. More precisely, we derive the Ord instance here: https://github.com/gibbon-compiler/gibbon/blob/2f97ff82f31ba23dc9f9778a7c4ddee52d1ab2ae/gibbon-compiler/src/Gibbon/L2/Syntax.hs#L203 The deriving code in GHC also has not (meaningfully) changed in decades.

Reply:
yeah sure that works for me. Thank you.